KIC Proxy Request guides (#879)
Kong/developer.konghq.com: app/_how-tos/kic-proxy-udp.md

---
title: Proxy UDP traffic by port
description: "Route UDP requests to services in your cluster using UDPRoute or UDPIngress"
content_type: how_to
related_resources:
  - text: All KIC documentation
    url: /index/kubernetes-ingress-controller/

permalink: /kubernetes-ingress-controller/routing/udp-by-port/
breadcrumbs:
  - /kubernetes-ingress-controller/
  - index: kubernetes-ingress-controller
    section: Routing

products:
  - kic

works_on:
  - on-prem
  - konnect

entities: []

tldr:
  q: How do I route UDP traffic with {{ site.kic_product_name }}?

>   a: Create a `UDPRoute` or `UDPIngress` resource, which will then be converted in to a {{ site.base_gateway }} Service and Route

Reviewer: ```suggestion
  a: Create a `UDPRoute` or `UDPIngress` resource, which will then be converted in to a [{{ site.base_gateway }} Service](/gateway/entities/service/) and [Route](/gateway/entities/route/).
```

prereqs:
  kubernetes:
    gateway_api: experimental
  entities:
    services:
      - udp-service

cleanup:
  inline:
    - title: Uninstall KIC from your cluster
      include_content: cleanup/products/kic
      icon_url: /assets/icons/kic.svg
---

## Add UDP listens

> {{site.base_gateway}} does not include any UDP listen configuration by default. To expose UDP listens, update the Deployment’s environment variables and port configuration.

Reviewer: ```suggestion
{{site.base_gateway}} doesn't include any UDP listen configuration by default. To expose UDP listens, update the Deployment’s environment variables and port configuration.
```

1. Set the `KONG_STREAM_LISTEN` environment variable and expose port `9999` in the Deployment:

    ```bash
    kubectl patch deploy -n kong kong-gateway --patch '{
      "spec": {
        "template": {
          "spec": {
            "containers": [
              {
                "name": "proxy",
                "env": [
                  {
                    "name": "KONG_STREAM_LISTEN",
                    "value": "0.0.0.0:9999 udp"
                  }
                ],
                "ports": [
                  {
                    "containerPort": 9999,
                    "name": "stream9999",
                    "protocol": "UDP"
                  }
                ]
              }
            ]
          }
        }
      }
    }'
    ```

> 1.  Update the proxy Service to indicate the new ports.

Reviewer: ```suggestion
1.  Update the proxy Service to indicate the new ports:
```

```bash
kubectl patch -n kong --type=json gateway kong -p='[
    {
        "op":"add",
        "path":"/spec/listeners/-",
        "value":{
            "name":"stream9999",
            "port":9999,
            "protocol":"UDP",
            "allowedRoutes": {
                "namespaces": {
                  "from": "All"
                }
            }
        }
    }
]'
```

Next, create a `UDPRoute`:

```bash
echo "apiVersion: gateway.networking.k8s.io/v1alpha2
kind: UDPRoute
metadata:
  name: tftp
  namespace: kong
spec:
  parentRefs:
  - name: kong
    namespace: kong
  rules:
  - backendRefs:
    - name: tftp
      port: 9999
" | kubectl apply -f -
```

{% endnavtab %}
{% navtab "Ingress" %}

```bash
echo "apiVersion: configuration.konghq.com/v1beta1
kind: UDPIngress
metadata:
  name: tftp
  namespace: kong
  annotations:
    kubernetes.io/ingress.class: kong
spec:
  rules:
  - backend:
      serviceName: tftp
      servicePort: 9999
    port: 9999
" | kubectl apply -f -
```

{% endnavtab %}
{% endnavtabs %}

This configuration routes traffic to UDP port `9999` on the
{{site.base_gateway}} proxy to port `9999` on the TFTP test server.

## Validate your configuration

> Send a TFTP request through the proxy.

Reviewer: ```suggestion
Send a TFTP request through the proxy:
```
